toolchain/sem_ir/dump.cpp

```diff
@@ -0,0 +1,244 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef NDEBUG
+
+#include "toolchain/sem_ir/dump.h"
+
+#include "common/ostream.h"
+#include "toolchain/sem_ir/stringify_type.h"
+
+namespace Carbon::SemIR {
+
+static auto DumpIfValid(const File& file, NameId name_id) -> void {
+  if (name_id.is_valid()) {
+    llvm::errs() << " `" << file.names().GetFormatted(name_id) << "`";
+  }
+}
+
+static auto DumpNoNewline(const File& file, ConstantId const_id) -> void {
+  llvm::errs() << const_id;
+  if (const_id.is_symbolic()) {
+    llvm::errs() << ": "
+                 << file.constant_values().GetSymbolicConstant(const_id);
+  } else if (const_id.is_template()) {
+    llvm::errs() << ": "
+                 << file.insts().Get(
+                        file.constant_values().GetInstId(const_id));
+  }
+}
+
+static auto DumpNoNewline(const File& file, InstId inst_id) -> void {
+  llvm::errs() << inst_id;
+  if (inst_id.is_valid()) {
+    llvm::errs() << ": " << file.insts().Get(inst_id);
+  }
+}
+
+static auto DumpNoNewline(const File& file, InterfaceId interface_id) -> void {
+  llvm::errs() << interface_id;
+  if (interface_id.is_valid()) {
+    auto interface = file.interfaces().Get(interface_id);
+    llvm::errs() << ": " << interface;
+    DumpIfValid(file, interface.name_id);
+  }
+}
+
+static auto DumpNoNewline(const File& file, SpecificId specific_id) -> void {
+  llvm::errs() << specific_id;
+  if (specific_id.is_valid()) {
+    llvm::errs() << ": " << file.specifics().Get(specific_id);
+  }
+}
+
+LLVM_DUMP_METHOD auto Dump(const File& file, ClassId class_id) -> void {
+  llvm::errs() << class_id;
+  if (class_id.is_valid()) {
+    auto class_obj = file.classes().Get(class_id);
+    llvm::errs() << ": " << class_obj;
+    DumpIfValid(file, class_obj.name_id);
+  }
+  llvm::errs() << '\n';
+}
+
+LLVM_DUMP_METHOD auto Dump(const File& file, ConstantId const_id) -> void {
+  DumpNoNewline(file, const_id);
+  llvm::errs() << '\n';
+}
+
+LLVM_DUMP_METHOD auto Dump(const File& file, EntityNameId entity_name_id)
+    -> void {
+  llvm::errs() << entity_name_id;
```

Comment:
I would suggest for each of these to write the type being dumped in the output too, see for example https://github.com/carbon-language/carbon-lang/pull/4747/files?diff=unified&w=0#diff-e690edf9337d3bd7e874588bc9d865536d6c69f2b7e8db79a9ef3aa9fd4b6edbL33-L42

So it could be something like `EntityNameId(...the stuff in here...)`

Reply:
I believe the print functions for each of the `Id` classes already includes its type: https://github.com/carbon-language/carbon-lang/blob/89df77707b6009abbc7687ab7e990e72f0b552e9/toolchain/base/index_base.h#L52